tf_encrypted/protocol/pond/pond.py

```diff
@@ -439,6 +439,119 @@ def helper(v: tf.Tensor) -> "PondPublicTensor":
+
+    Example:
+      ```
+      @tfe.local_computation('input-provider')
+      def provide_input():
+        return tf.random.normal((3, 3))
+
+      @tfe.local_computation
+      def receive_output(logits):
+        return tf.print(tf.argmax(logits, axis=-1))
+
+      x = provide_input()
+      y = model(x)
+      receive_op = receive_output(y, player_name='output-receiver')
+      with tfe.Session():
+        sess.run(receive_op)
+      ```
+
+    Arguments:
+      player_name: Name of the player who should execute the function.
+      kwargs: Keyword arguments to use when encoding or encrypting
+        inputs/outputs to compute_func: see tfe.define_local_computation for
+        details.
+
+    Returns:
+      The compute_func, but decorated for remote execution.
+    """
+    if callable(player_name):
+      # The user has called us as a standard decorator:
+      #
+      # @tfe.local_computation
+      # def provide_input():
+      #   return tf.zeros((2, 2))
+      actual_compute_func = player_name
+      player_name = None
+    else:
+      # The user has called us as a function, maybe with non-default args:
+      #
+      # @tfe.local_computation('input-provider')
+      # def provide_input():
+      #   return tf.zeros((2, 2))
+      actual_compute_func = None
+
+    def decorator(compute_func):
+
+      @wraps(compute_func)
+      def compute_func_wrapper(*compute_func_args, **compute_func_kwargs):
+
+        # Assumer user has passed player_name to decorator. If not, try to recover.
+        actual_player_name = player_name
+        if actual_player_name is None:
+          # Maybe user has passed player_name to compute_func as a kwarg
+          actual_player_name = compute_func_kwargs.get("player_name", None)
+        if actual_player_name is None:
+          # Assume compute_func is a method and its instance has some attribute
+          # 'player_name'
+          if compute_func_args:
+            parent_instance = compute_func_args[0]
+            actual_player_name = getattr(parent_instance, 'player_name', None)
+        if actual_player_name is None:
+          # Fallback to error
+          raise ValueError("'player_name' not provided. Please provide "
+                           "'player_name' as a keyword argument to this "
+                           "function, or as an argument to the "
+                           "tfe.local_computation decorator.")
+
+        def result_guard(*args, **kwargs):
+          output = compute_func(*args, **kwargs)
+
+          # wrap in tf.group to prevent sending back any tensors (which might
+          # hence be leaked)
+          is_op = tf.contrib.framework.nest.flatten(
+              tf.contrib.framework.nest.map_structure(
+                  lambda x: isinstance(x, tf.Operation),
+                  output,
+              )
+          )
+          if all(is_op):
+            return tf.group(output)
+
+          return output
```

Comment:
this bypasses the security precaution even if only a single value is not an op. until we need it, it might be better to simply assert that only `tf.Operation`s are returned, and always group